### DISCUSSION TITLE
How does Meshoptimizer classify vertices before simplification?

### DISCUSSION THREAD

**Discussion by daniilvinn** [zeux/meshoptimizer#693] [Q&A]
Hello!
I develop Nanite-like Virtual Geometry system, and I need to build a hiererachy of meshlets. On each step I group meshlets, merge their indices and simplify with *locked borders*.

However, I am struggling to implement it properly - for some reason, simplifier gets stuck *really* often.

I found out that there are 5 kinds of vertices - complex, manifold, seam, border and locked vertices. With locked borders on, every border vertex becomes locked vertex.

Also I found out that when I try to simplify Standford's Bunny mesh (fully connected mesh), at last simplification passes I get only 1 meshlet group which consist of 2 meshlets. After debugging, I found out that vertex classificator still classifies some vertices as border ones (meaning that they are locked). It is unexpected behaviour from my point of view - how can it find borders even though that mesh is fully connected and I have only 1 group (entire mesh)? Shouldn't borders be attribute seams?
This behaviour leads to simplifier getting stuck. I also tested on Happy Buddha's mesh and it gets stuck much earlier, so I can't really simplify it properly. it just fails to simplify -> I reregister meshlets which failed to simplify -> it repeats for next pass.

Also I don't understand what "complex" and "manifold" vertex kinds are and how do "border" and "seam" kinds differ?

**Comment by zeux** [comment]
The likely cause is that when merging meshlets, you need to:
1. Merge neighboring meshlets as much as possible so that the border between meshlets is not “open” (triangles on both sides of the border are part of the merged structure)
2. Ensure that the border between meshlets has the same vertex ids for triangles on both sides.

If most vertices are classified as border, this is unusual but could mean that your indices are incorrect - not pointing to vertices with correct positions perhaps - and the simplifier can’t find adjacent triangles because of this. This also would happen if you don’t do 1 well; if you don’t do 2 well you will get seam or locked vertices.

https://github.com/zeux/meshoptimizer/blob/b8843d9112d39b20bb3abc0da0e54d8a2b3b23a6/src/simplifier.cpp#L208 describes the classification; Complex is currently not used. Manifold describes a regular vertex that isn’t on a seam, a border or any difficult topological configuration that is classified as Locked.

**Reply by daniilvinn** [reply]
Hello! I didn't mean that "most of vertices are border" I meant that it is in fact incorrect that some vertices are classified as border, and not seam. Also, from your quote:
> Ensure that the border between meshlets has the same **vertex ids** for triangles on both sides.

A "vertex id" is an index, right?

**Reply by zeux** [reply]
Yeah, index.

When a vertex is classified as a border, it means that two of the edges the vertex belongs to are open - open means there’s just one triangle with this edge, and not two like there normally would be. This analysis includes triangles with a different vertex index but same position; if the edge was closed from the opposite side by a triangle with same positions but different indices, the vertex would be classified as seam.

In Nanite style simplification, after you merge the clusters, the expectation is:
- There are going to be exterior vertices that will be on the border of the merged structure. They will need to be classified as border. Note that from the perspective of the entire mesh these aren’t on a border - there are triangles on another side - but the simplification process works on a subset of the mesh 
- The fewer clusters you merge, the more border vertices you will get. For example it’s a mistake to merge just two. I think Nanite merges something like 32+ clusters but it’s been a while. Also as mentioned above, you need to merge adjacent clusters to ensure there’s a minimum number of the border vertices.
- Other vertices should normally be manifold most of the time. For example the shared border between merged clusters should be classified as manifold if you make sure the vertices have the same index. If everything is done correctly, you should generally only get seam vertices for attribute discontinuities - like a UV seam in the source mesh. A sphere without any texture coordinates should only have manifold vertices and border vertices during Nanite style processing for any cluster group you work on.

**Reply by daniilvinn** [reply]
Thanks for information!
Small notes: I *do not* group clusters by 2, I group them by 4 and METIS library outputs groups with not-fixed size - even if I have 16 meshlets and *num_partitions* is 4, I may get 2 groups with 5 meshlets and 2 groups with 3 meshlets. 

At last LOD levels I get a single group of 2 meshlets just because other were simplified out, so entire LOD level consists of just 1 group which has 2 meshlets. Hence it means that if mesh is fully connected and it was simplified to a single group, there should be no border vertices *at all*, because entire mesh consists of just 1 group (which was merged from, say, 2 meshlets).

Thanks for explanation of seam/border vertex kind difference. This is exactly what I'm debugging right now - for some reason, vertices are classified as border, even though they should to be classified as seam or manifold (**every** edge is closed - only a single group, as I mentioned above).

I understand that on lower LOD levels, where mesh is highlier detailed, I may indeed get border (thus locked) vertices. But at last levels, where entire LOD consists of a single group and source mesh is fully connected there should be no border vertices, as I understand <- this is the problem.
